plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/SQLEditorOutputViewer.java

```diff
@@ -18,123 +18,213 @@
-        return text;
+        TextEditorUtils.enableHostEditorKeyBindingsSupport(site, filterText);
     }
 
-    void refreshStyles() {
-        ITheme currentTheme = PlatformUI.getWorkbench().getThemeManager().getCurrentTheme();
-        Font outputFont = currentTheme.getFontRegistry().get(SQLConstants.CONFIG_FONT_OUTPUT);
-        if (outputFont != null) {
-            this.text.setFont(outputFont);
+    @Override
+    public void println(@Nullable DBCOutputSeverity severity, @Nullable String message) {
+        if (message == null) {
+            return;
+        }
+        if (severity == null || severities.contains(severity)) {
+            viewer.getOutputWriter().println(message);
+        }
+        records.offer(new OutputRecord(severity, message));
+        if (records.size() > MAX_RECORDS) {
+            records.pop();
         }
-        this.text.setForeground(currentTheme.getColorRegistry().get(SQLConstants.CONFIG_COLOR_TEXT));
-        this.text.setBackground(currentTheme.getColorRegistry().get(SQLConstants.CONFIG_COLOR_BACKGROUND));
     }
 
-    void print(String out)
-    {
-        text.append(out);
+    @Override
+    public void flush() {
+        viewer.getOutputWriter().flush();
     }
 
-    void println(String out)
-    {
-        print(out + "\n");
+    @NotNull
+    public SQLEditorOutputConsoleViewer getViewer() {
+        return viewer;
     }
 
-    void scrollToEnd()
-    {
-        text.setTopIndex(text.getLineCount() - 1);
+    public boolean isHasNewOutput() {
+        return viewer.isHasNewOutput();
     }
 
-    public PrintWriter getOutputWriter() {
-        return writer;
+    public void refreshStyles() {
+        viewer.refreshStyles();
     }
 
-    public boolean isHasNewOutput() {
-        return hasNewOutput;
+    public void resetNewOutput() {
+        viewer.resetNewOutput();
+    }
+
+    public void clearOutput() {
+        viewer.clearOutput();
+    }
+
+    public void setExecutionContext(@NotNull DBCExecutionContext executionContext) {
+        if (this.executionContext != executionContext) {
+            this.executionContext = executionContext;
+
+            clearOutput();
+            severities.clear();
+
+            final DBPDataSource dataSource = executionContext.getDataSource();
+            final DBCServerOutputReader reader = DBUtils.getAdapter(DBCServerOutputReader.class, dataSource);
+
+            if (reader instanceof DBCServerOutputReaderExt) {
+                final DBCServerOutputReaderExt readerExt = (DBCServerOutputReaderExt) reader;
+                final DBCOutputSeverity[] supportedSeverities = readerExt.getSupportedSeverities(executionContext);
+
+                severities.addAll(List.of(supportedSeverities));
+
+                UIUtils.setControlVisible(filterToolbar.getControl(), supportedSeverities.length > 0);
+            } else {
+                UIUtils.setControlVisible(filterToolbar.getControl(), false);
+            }
+        }
```

Comment:
Pretty complex logic for set method, maybe `updateExecutionContext` with the proper javadoc?

Reply:
Nah?